### PR TITLE
ignore_comments_in_yaml_file_comparison

### PR DESCRIPTION
**Pull request by mikaelarguedas**
this adds a parameter to ignore comments when comparing distribution / index files. this should allow https://github.com/ros/rosdistro/pull/14517 to be merged as  is

### CODE REVIEW

**Comment by mikepurvis** [comment]
Strongly in support of this direction— however, note that without a wholesale switch to [ruamel.yaml](https://pypi.python.org/pypi/ruamel.yaml), other tools are still going to mangle/drop comments in `distribution.yaml` (eg Bloom PR creation; anything that uses the rosdistro python library).

**Comment by mikaelarguedas** [comment]
> a wholesale switch to ruamel.yaml

Thanks for the pointer, I was looking for a parser implementation supporting comments!

> other tools are still going to mangle/drop comments in distribution.yaml (eg Bloom PR creation; anything that uses the rosdistro python library).

That is a very good point.

I don't know if @ros-infrastructure/ros_team would be willing to change the parser completely given that this will also impact any code interacting with distribution files. Maybe the simplest is to just remove the EOL distros instead of commenting them out as suggested [here](https://github.com/ros/rosdistro/pull/14517#issuecomment-294379292). Thoughts?

File: src/rosdistro/verify.py [hunk 1]
```diff
@@ -128,17 +128,29 @@ def _check_files_identical(index, dist_name, loader_function, yaml_url, file_typ
...
+        import re
+        yaml_lines_to_remove = []
+        yaml_header = _yaml_header_lines(file_type, dist_file_data['version'])
+        for yaml_line in yaml_lines:
+            if yaml_line in yaml_header:
+                continue
+            elif re.match(r'(^#| +#)', yaml_line):
+                yaml_lines_to_remove.append(yaml_line)
+        for yaml_line_to_remove in yaml_lines_to_remove:
+            yaml_lines.remove(yaml_line_to_remove)
```
Comment (dirk-thomas) [inline src/rosdistro/verify.py]: Could this logic be simplified with a list comprehension? E.g. like this (untested):

```
yaml_lines = [
    line for line in yaml_lines
    if line not in yaml_header and not line.lstrip().startswith('#')]
```

**Comment by dirk-thomas** [comment]
I think the only tool in our pipeline writing the distribution files is `bloom` (which uses `rosdistro` for this internally?). Those would need to be updated to use a parser which keeps comments. I am not sure if it is worth the effort.

File: src/rosdistro/verify.py (hunk 1 above)
Comment (mikaelarguedas) [inline src/rosdistro/verify.py]: good point, now that I simplified the regex to not process the header but only the comment lstrip will totally do the job :+1:
fixed in https://github.com/ros-infrastructure/rosdistro/pull/98/commits/e1e84a0282445b8ee74099b9aa9c4b0b374daaf5

**Comment by mikaelarguedas** [comment]
I don't think I'll put the effort in changing parser in the near future. Mostly because I think that relying on python core modules is preferred vs using third party packages. I modified https://github.com/ros/rosdistro/pull/14517 to remove the EOL distros rather than commenting them out. @ros-infrastructure/ros_team feel free to either close this or keep it open for visibility if we ever want to build on top.

**Comment by dirk-thomas** [comment]
Since this would require the returned data structures to fundamentally change I don't think this will ever happen. So I will close it.

**Pull request closed without merging**